lib/cusparse/interfaces.jl

```diff
@@ -127,6 +127,73 @@ function Base.:(-)(A::CuSparseMatrix, B::CuSparseMatrixCSR)
     return geam(one(eltype(A)), csrA, -one(eltype(A)), B, 'O')
 end
 
+function Base.reshape(A::CuSparseMatrixCOO{T,M}, dims::NTuple{N,Int}) where {T,N,M}
+    nrows, ncols = size(A)
+    flat_indices = nrows .* (A.colInd .- 1) .+ A.rowInd .- 1
+    new_col, new_row = div.(flat_indices, dims[1]) .+ 1, rem.(flat_indices, dims[1]) .+ 1
+    sparse(new_row, new_col, A.nzVal, dims[1], length(dims) == 1 ? 1 : dims[2], fmt = :coo)
+end
+
+function LinearAlgebra.mul!(Y::CuSparseMatrixCSR{T,M}, A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}) where {T,M}
+    CUSPARSE.version() < v"11.5.1" && throw(ErrorException("This operation is not supported by the current CUDA version."))
+    mm!('N', 'N', one(T), A, B, zero(T), Y, 'O')
+end
```

Comment:
```suggestion
function LinearAlgebra.mul!(Y::CuSparseMatrixCSR{T,M}, A::CuSparseMatrixCSR{T,M}, B::CuSparseMatrixCSR{T,M}, alpha::Number, beta::Number) where {T,M}
    CUSPARSE.version() < v"11.5.1" && throw(ErrorException("This operation is not supported by the current CUDA version."))
    gemm!('N', 'N', alpha, A, B, beta, Y, 'O')
end
```

Reply:
They also use `gemm!` for sparse matrix - sparse matrix products in CUSPARSE.
`mm!` is used for sparse matrix - dense matrix products.
I updated the interface a few days ago to fix some issues with this routine (#1613).

Reply:
Ok you are right. Bu why does it work with two sparse matrices, and it returns the correct sparse matrix? Anyway, I will change `mm!` to `gemm!`.

Reply:
I suppose that you didn't rebase your branch with `master`.

Reply:
I updated my branch, but I went into an error

```julia
using CUDA
using CUDA.CUSPARSE
CUDA.allowscalar(false)
using LinearAlgebra
using SparseArrays

A = sprand(ComplexF32, 100, 100, 0.1)
A_gpu = CuSparseMatrixCSR(A)

Y = gemm('N', 'N', one(eltype(A_gpu)), A_gpu, A_gpu, 'O');
Y
Array(Y) ≈ A * A
```
with `gemm` works. But if I want to use the in-place function `gemm!` it fails

```julia
Y = similar(A_gpu)
gemm!('N', 'N', one(eltype(A_gpu)), A_gpu, A_gpu, zero(eltype(A_gpu)), Y, 'O')
Array(Y) ≈ A * A
```

It gives me the error

```julia
ERROR: CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA ~/.julia/dev/CUDA/lib/cudadrv/error.jl:89
  [2] macro expansion
    @ ~/.julia/dev/CUDA/lib/cudadrv/error.jl:97 [inlined]
  [3] cuMemAllocAsync
    @ ~/.julia/dev/CUDA/lib/utils/call.jl:26 [inlined]
  [4] alloc(::Type{CUDA.Mem.DeviceBuffer}, bytesize::Int64; async::Bool, stream::CuStream, pool::Nothing)
    @ CUDA.Mem ~/.julia/dev/CUDA/lib/cudadrv/memory.jl:83
  [5] actual_alloc(bytes::Int64; async::Bool, stream::CuStream)
    @ CUDA ~/.julia/dev/CUDA/src/pool.jl:39
  [6] macro expansion
    @ ~/.julia/dev/CUDA/src/pool.jl:222 [inlined]
  [7] macro expansion
    @ ./timing.jl:382 [inlined]
  [8] #_alloc#166
    @ ~/.julia/dev/CUDA/src/pool.jl:303 [inlined]
  [9] #alloc#165
    @ ~/.julia/dev/CUDA/src/pool.jl:289 [inlined]
 [10] alloc
    @ ~/.julia/dev/CUDA/src/pool.jl:283 [inlined]
 [11] CuArray
    @ ~/.julia/dev/CUDA/src/array.jl:85 [inlined]
 [12] CuArray
    @ ~/.julia/dev/CUDA/src/array.jl:168 [inlined]
 [13] CuArray
    @ ~/.julia/dev/CUDA/src/array.jl:175 [inlined]
 [14] with_workspace(f::CUDA.CUSPARSE.var"#2349#2352"{ComplexF32, Char, Char, ComplexF32, ComplexF32, CUDA.CUSPARSE.cusparseSpGEMMAlg_t, CUDA.CUSPARSE.CuSpGEMMDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor}, eltyp::Type{UInt8}, size::CUDA.CUSPARSE.var"#buffer1Size#2351"{ComplexF32, Char, Char, ComplexF32, ComplexF32, CUDA.CUSPARSE.cusparseSpGEMMAlg_t, CUDA.CUSPARSE.CuSpGEMMDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor, CUDA.CUSPARSE.CuSparseMatrixDescriptor}, fallback::Nothing; keep::Bool)
    @ CUDA.APIUtils ~/.julia/dev/CUDA/lib/utils/call.jl:65
 [15] #with_workspace#1
    @ ~/.julia/dev/CUDA/lib/utils/call.jl:53 [inlined]
 [16] gemm!(transa::Char, transb::Char, alpha::ComplexF32, A::CuSparseMatrixCSR{ComplexF32, Int32}, B::CuSparseMatrixCSR{ComplexF32, Int32}, beta::ComplexF32, C::CuSparseMatrixCSR{ComplexF32, Int32}, index::Char, algo::CUDA.CUSPARSE.cusparseSpGEMMAlg_t)
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/cusparse/generic.jl:258
 [17] gemm!(transa::Char, transb::Char, alpha::ComplexF32, A::CuSparseMatrixCSR{ComplexF32, Int32}, B::CuSparseMatrixCSR{ComplexF32, Int32}, beta::ComplexF32, C::CuSparseMatrixCSR{ComplexF32, Int32}, index::Char)
    @ CUDA.CUSPARSE ~/.julia/dev/CUDA/lib/cusparse/generic.jl:236
 [18] top-level scope
    @ Untitled-1:256
```

I think this is related to what you said before, that Y must to have the same sparsity of the product. Is it correct? Just only to be sure that my branch works.

PS: I was asking when it is useful to use the in-place version since I have to know the sparsity of the product? And how can I know the sparsity pattern of the product without calculating it?

Reply:
Yes, if you want to use the in-place `gemm!`, `C` must have the same sparsity pattern of the product `AB`.
If you don't know the sparsity pattern, `gemm` will allocated a new `CuSparseMatrixCSR` for you.

The previous version of SpGEMM (named `mm!`) was a hack.
It took as input a CuSparseMatrixCSR but allocated new `CuVector`s for `rowvals` and `nzvals` and updated the pointers.
It was as expensive as the current `gemm` and was not working if `beta != 0`.

It can be relevant to use the in-place if you want to solve multiple linear systems with normal equations for instance.
If you want to solve a linear problem `min c'x  subject to Ax=b, x>=0` with an interior point method you will need to solve multiple linear systems with `A'DA` where D is a diagonal matrix that is updated at each iteration of the optimization method.
The first time you call `gemm` but after you could just use `gemm!` because the sparsity pattern will be the same.

Reply:
Ok, very clear. Thank you.

I think I updated all the functions now.